File: src/main/scala/booklet/pure/model/BookData.scala
package booklet.pure.model

import zio.json.{DeriveJsonEncoder, JsonEncoder}

case class BookData(
    isbn: Option[Isbn],
    author: Option[Author],
    title: Option[Title],
    subtitle: Option[Subtitle],
    thumbnail: Option[String],
    smallThumbnail: Option[String],
    userId: Option[String] // P5628
)

object BookData {
  implicit val encoder: JsonEncoder[BookData] = DeriveJsonEncoder.gen[BookData]

  private val isbn = "isbn"
  private val authr = "author"
  private val ttle = "title"

  def completeFromHttpQuery(qry: Map[String, String]): Option[BookData] = for {
    isbn <- qry.get(isbn)
    author <- qry.get(authr)
    title <- qry.get(ttle)
  } yield BookData(
    isbn = Some(Isbn(isbn)),
    author = Some(Author(author)),
    title = Some(Title(title)),
    subtitle = None,
    thumbnail = None,
    smallThumbnail = None,
    userId = None // P5628
  )

  def partialFromHttpQuery(qry: Map[String, String]): BookData = BookData(
    isbn = qry.get(isbn).map(Isbn.apply),
    author = qry.get(authr).map(Author.apply),
    title = qry.get(ttle).map(Title.apply),
    subtitle = None,
    thumbnail = None,
    smallThumbnail = None,
    userId = None // P5628
  )
}

File: src/main/ts/booklet/pure/model/BookData.ts
import { Isbn } from './Isbn';
import { Author } from './Author';
import { Title } from './Title';
import { Subtitle } from './Subtitle';

export class BookData {
  constructor(
    public isbn: Isbn | null,
    public author: Author | null,
    public title: Title | null,
    public subtitle: Subtitle | null,
    public thumbnail: string | null,
    public smallThumbnail: string | null,
    public userId: string | null
  ) {}

  static completeFromHttpQuery(qry: { [key: string]: string }): BookData | null {
    const isbn = qry['isbn'];
    const author = qry['author'];
    const title = qry['title'];

    if (isbn && author && title) {
      return new BookData(
        new Isbn(isbn),
        new Author(author),
        new Title(title),
        null,
        null,
        null,
        null
      );
    }

    return null;
  }

  static partialFromHttpQuery(qry: { [key: string]: string }): BookData {
    return new BookData(
      qry['isbn'] ? new Isbn(qry['isbn']) : null,
      qry['author'] ? new Author(qry['author']) : null,
      qry['title'] ? new Title(qry['title']) : null,
      null,
      null,
      null,
      null
    );
  }
}
